### PR TITLE
Nashwan

### PR DESCRIPTION
**Pull request by NNSSA**

### CODE REVIEW

File: tests/unit/interpolation_test.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def binary_search_test():
 
     # Test the case with needle outside the haystack range
     assert binary_find(haystack, -1.) == (0, 0)
-    assert binary_find(haystack, 990.) == (98, 98)
+    assert binary_find(haystack, 98.) == (98, 98)
```
Comment (ckald) [inline tests/unit/interpolation_test.py]: You cannot arbitrarily change a test case such that it "works" :) 
The argument has to be 990 here, because for some reason for the code to function, the binary search has to be incorrect in this corner case. We are looking into this now. 

To be fair, I should have left a comment and probably marked the test case a "expected fail".

File: Makefile [hunk 2]
```diff
@@ -3,10 +3,10 @@ all: clean compile test
 compile: compile3
 
 compile3:
-	cd interactions/four_particle/cpp && c++ -fopenmp -O3 -shared -std=c++11 -I pybind11/include `python3-config --cflags --ldflags` integral.cpp -o integral.so
+	cd interactions/four_particle/cpp && c++ -fopenmp -O3 -shared -std=c++11 -fPIC -I pybind11/include `python3.5-config --cflags --ldflags` integral.cpp -o integral.so
```
[the remark this answers is not in the document]
Comment (NNSSA) [reply Makefile]: PIC = Position Independent Code.
Some say that with this flag the generated code is not dependent on being located at a specific address in memory in order to work.

Error disappeared when including this flag.

File: evolution.py [hunk 3]
```diff
@@ -130,11 +131,13 @@ def evolve(self, T_final, export=True, init_time=True):
         if export:  # and not interrupted:
             self.export()
 
-        return self.data
+#        return self.data
```
[the remark this answers is not in the document]
Comment (NNSSA) [reply evolution.py]: Not every function needs to return something. Especially when you won't do anything with the returned object.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (ckald) [reply Makefile]: Adding this flag does not hurt my compilation, but the internet says it's a signature of an old compiler... What's your version? I have GCC 7.2.0

[the remark this answers is not in the document]
Comment (NNSSA) [reply Makefile]: I have 4.8.5 . 
Will update it and see if error disappears.

File: evolution.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ckald) [reply evolution.py]: In Python every function returns something — even if it's `None`. I think you're obsessing over irrelevant things with these returns. I added this line because I predicted that one day I might want to do something with `Universe.data` after running a simulation. I don't use it now and I can access it in other ways — that's true, but this change literally does nothing for anybody :)

File: particles/__init__.py [hunk 4]
```diff
@@ -220,8 +220,11 @@ def update(self, force_print=False):
             'energy_density': self.energy_density
         })
 
+        #if force_print or self.regime != oldregime or self.in_equilibrium != oldeq:
+        #    print(self)
+
         if force_print or self.regime != oldregime or self.in_equilibrium != oldeq:
-            print(self)
+           print("\n"+ "\t" * 2 + "%s decoupled at T_dec = %.2f MeV \n" % (self.name, self.decoupling_temperature / UNITS.MeV) + ("\t" * 2 + "-" * 50))
```
[the remark this answers is not in the document]
Comment (ckald) [reply particles/__init__.py]: Hey, also there's a 3-space indentation. Do you mind installing something Python-specific instead of gedit? I use Sublime Text 3, but PyCharm might be even better but heavier

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (NNSSA) [reply Makefile]: Compiled again with v7.2.0 , but I still get same error.

[the remark this answers is not in the document]
Comment (ckald) [reply Makefile]: Then it's time to take a look at the error :)

[the remark this answers is not in the document]
Comment (NNSSA) [reply Makefile]: I get the following error:

/usr/bin/ld: /tmp/ccIJs4Lw.o: relocation R_X86_64_32S against `_ZNSs4_Rep20_S_empty_rep_storageE' can not be used when making a shared object; recompile with -fPIC
/tmp/ccIJs4Lw.o: error adding symbols: Bad value

Each time the file it refers to (in this case ccIJs4Lw.o), is different.

File: particles/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NNSSA) [reply particles/__init__.py]: That one was my fault. Standard indentation is set to 4 spaces.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (ckald) [reply Makefile]: Weird, but okay. I don't know what to do with it — let's leave your new flag.

[the remark this answers is not in the document]
Comment (ckald) [reply Makefile]: The filename is different because it's a temporary file

File: __init__.py [hunk 5]
```diff
@@ -33,10 +41,10 @@
 To run some script in the context of the code, one should temporarily alter the `PYTHONPATH`\
 environment variable:
 
-    PYTHONPATH=. python3 some_script.py
+    PYTHONPATH=. python3.5 some_script.py
```
Comment (ckald) [inline __init__.py]: What about this 3.5?

File: __init__.py [hunk 6]
```diff
@@ -33,10 +41,10 @@
...
 
 For example, to run the cosmic neutrino background temperature test:
 
-    PYTHONPATH=. python3 tests/cosmic_neutrino_temperature
+    PYTHONPATH=. python3.5 tests/cosmic_neutrino_temperature
```
Comment (ckald) [inline __init__.py]: What about this 3.5?

File: tests/unit/interpolation_test.py [hunk 7]
```diff
@@ -59,7 +59,7 @@ def binary_search_test():
 
     # Test the case with needle outside the haystack range
     assert binary_find(haystack, -1.) == (0, 0)
-    assert binary_find(haystack, 990.) == (99, 99)
+    assert binary_find(haystack, 99.) == (99, 99)
```
Comment (ckald) [inline tests/unit/interpolation_test.py]: Remove these changes

File: __init__.py (hunk 6 above)
Comment (NNSSA) [reply __init__.py]: The command python3 won't necessarily run python3.5

Comment (ckald) [reply __init__.py]: And `python3.5` definitely won't run Python 3.6. And there's no such command on a system where Python 3.6 is the default interpreter

Comment (NNSSA) [reply __init__.py]: I added "using for example Python 3.5" in the current one.
